File: minimum-falling-path-sum/minimum-falling-path-sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int min_path=INT_MAX;
        std::map<std::string,int> m1;
        for(int itr1=0;itr1<matrix[0].size();itr1++){
            min_path=std::min(min_path,min_path_sum(matrix,0,itr1,m1));
        }
        return min_path;
    }
    
    int min_path_sum(std::vector<vector<int>>& matrix,int src_row,int src_col,std::map<std::string,int>& m1){
        if(src_row<0 || src_col<0 || src_row>=matrix.size() || src_col>=matrix[0].size()){
            return INT_MAX;
        }
        if(src_row==matrix.size()-1){
            return matrix[src_row][src_col];
        }
        std::string key=to_string(src_row)+"_"+to_string(src_col);
        if(m1[key]!=0){
            return m1[key];
        }
        int diag_right=min_path_sum(matrix,src_row+1,src_col+1,m1);
        int diag_left=min_path_sum(matrix,src_row+1,src_col-1,m1);
        int down=min_path_sum(matrix,src_row+1,src_col,m1);
        int res=std::min(diag_right,diag_left);
        res=std::min(res,down);
        m1[key]= res+matrix[src_row][src_col];
        return m1[key];
    }
};

File: 967-minimum-falling-path-sum/minimum-falling-path-sum.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        return minPathIterative(matrix);

        // int min = (int) 1e9;
        // int row = matrix.length-1;
        // int dp[][] = new int[row+1][matrix[0].length];
        // for (int i = 0; i < dp.length; i++) {
        //     Arrays.fill(dp[i], -1); 
        // }

        // for(int itr1=0; itr1<matrix[0].length;itr1++){
        //     min = Math.min(min,minPath(row,itr1,matrix,dp));
        // }
        // return min;
    }

    private int minPathIterative(int[][] mat){
        int row = mat.length;
        int col = mat[0].length;
        int dp[][] = new int[row][col];
        
        for(int itr1=0; itr1< col; itr1++){
            dp[0][itr1] = mat[0][itr1];
        }
        
        for(int itr1=1; itr1<row;itr1++){
            for(int itr2=0;itr2<col;itr2++){
                int left = (int) 1e9;
                int right = (int) 1e9;
                if(itr2>0)
                    left = mat[itr1][itr2] + dp[itr1-1][itr2-1];
                int down = mat[itr1][itr2] + dp[itr1-1][itr2];
                if(itr2+1<col)
                    right = mat[itr1][itr2]+ dp[itr1-1][itr2+1];
                dp[itr1][itr2] = Math.min(Math.min(left,down),right);
            }
        }
        int min = (int) 1e9;
        for(int itr1 =0;itr1<col;itr1++){
            min = Math.min(min,dp[row-1][itr1]);
        }
        return min;   
    }
    
    private int minPath(int row, int col, int[][] matrix,int dp[][]){
        if(col < 0 || col>= matrix[0].length){
            return (int) 1e9;
        }
        if(row == 0){
            return matrix[row][col];
        }
        if(dp[row][col]!=-1){
            return dp[row][col];
        }
        
        int left = matrix[row][col] + minPath(row-1,col-1,matrix,dp);
        int up = matrix[row][col] + minPath(row-1,col,matrix,dp);
        int right = matrix[row][col] + minPath(row-1,col+1,matrix,dp);
        dp[row][col] = Math.min(Math.min(left,up),right);
        return dp[row][col];
    }
}